### ISSUE TITLE
PCKE

### ISSUE THREAD

**Issue by Maesla** [cdmvision/authentication-unity#5]
Hi @cdmdeveloper 

I'm adding PCKE to your implementation
https://www.rfc-editor.org/rfc/rfc7636

It seems to work, but for making this work, I have to change several classes (AccessTokenRequest, AuthentitcationSession, AuthorizationCodeFlow, AuthorizationCodeRequest). 

These would be the major changes:

```csharp
        public string GetAuthorizationUrl()
        {
            // Generate new state.
            state = Guid.NewGuid().ToString("D");

            codeVerifier = GenerateRandomDataBase64url(32);
            string codeChallenge = Base64UrlEncodeNoPadding(Sha256Ascii(codeVerifier));
            const string codeChallengeMethod = "S256";

            var parameters = JsonHelper.ToDictionary(new AuthorizationCodeRequest()
            {
                clientId = configuration.clientId,
                redirectUri = configuration.redirectUri,
                scope = configuration.scope,
                state = state,
                codeChallenge = codeChallenge,
                codeChallengeMethod = codeChallengeMethod
            });

            return UrlBuilder.New(authorizationUrl).SetQueryParameters(parameters).ToString();
        }
```

```csharp
public virtual async Task<AccessTokenResponse> ExchangeCodeForAccessTokenAsync(string redirectUrl,
            CancellationToken cancellationToken = default)
        {
            var authorizationResponseUri = new Uri(redirectUrl);
            var query = HttpUtility.ParseQueryString(authorizationResponseUri.Query);

            // Is there any error?
            if (JsonHelper.TryGetFromNameValueCollection<AuthorizationCodeRequestError>(query, out var authorizationError))
                throw new AuthorizationCodeRequestException(authorizationError);

            if (!JsonHelper.TryGetFromNameValueCollection<AuthorizationCodeResponse>(query, out var authorizationResponse))
                throw new Exception("Authorization code could not get.");

            // Validate authorization response state.
            if (!string.IsNullOrEmpty(state) && state != authorizationResponse.state)
                throw new SecurityException($"Invalid state got: {authorizationResponse.state}");

            var parameters = JsonHelper.ToDictionary(new AccessTokenRequest()
            {
                code = authorizationResponse.code,
                clientId = configuration.clientId,
                clientSecret = configuration.clientSecret,
                redirectUri = configuration.redirectUri,
                codeVerifier = codeVerifier
            });

            UnityEngine.Debug.Assert(parameters != null);

            accessTokenResponse = 
                await GetAccessTokenInternalAsync(new FormUrlEncodedContent(parameters), cancellationToken);
            return accessTokenResponse;
        }
```

I'd like to add this functionality in a cleaner way, because I don't want to lose the possibility of using a clean grant code without PCKE

What do you think it could be the cleaner way to do it?

Maybe in the methods GetAuthorizationUrl you can have a virtual GetAuthorizationUrlParameters that can be overwritten in a  AuthorizationCodeFlowWithPCKE class
```csharp
        public string GetAuthorizationUrl()
        {
            // Generate new state.
            state = Guid.NewGuid().ToString("D");

            codeVerifier = GenerateRandomDataBase64url(32);
            string codeChallenge = Base64UrlEncodeNoPadding(Sha256Ascii(codeVerifier));
            const string codeChallengeMethod = "S256";

            var parameters = GetAuthorizationUrlParameters ();


            return UrlBuilder.New(authorizationUrl).SetQueryParameters(parameters).ToString();
        }

protected virtual Dictionary<string, string> GetAuthorizationUrlParameters ()
{
return JsonHelper.ToDictionary(new AuthorizationCodeRequest()
            {
                clientId = configuration.clientId,
                redirectUri = configuration.redirectUri,
                scope = configuration.scope,
                state = state,
                codeChallenge = codeChallenge,
                codeChallengeMethod = codeChallengeMethod
            });
}
```

Same for ExchangeCodeForAccessTokenAsync.

Thank you

**Comment by Maesla** [comment]
I've added a PR with my proposal 
#6 
Thank you

**Comment by ibrahimpenekli** [comment]
Thanks for contribution! I renamed `AuthorizationCodeFlowWithPCKE` to `AuthorizationCodeFlowWithPkce` for consistency fyi.. Released in 1.2.0
